feat: Implement structured Authorization header

jakubjanecek opened the pull request:
Part of #1524.

Commit:
feat: Implement structured Authorization header

Wosin commented on zio-http/src/main/scala/zio/http/model/headers/values/Authorization.scala:
```diff
@@ -0,0 +1,147 @@
+      cnonce: String,
+      nc: Int,
+      userhash: Boolean,
+      rest: Map[String, String],
```
IMO, we should rename that to something else than `rest` as it may misleading (especially can be similar to REST).

jakubjanecek replied:
I removed it altogether. I thought I would need it but it's not necessary.

Wosin commented on zio-http/src/main/scala/zio/http/model/headers/values/Authorization.scala:
```diff
@@ -0,0 +1,147 @@
+    var index                = 0
+    var params               = HashMap[String, String]()
+    var rest                 = HashMap[String, String]()
+    var currentKey: String   = null
```
Is there a way to get rid of those `var` and `nulls`?

jakubjanecek replied:
Sure there is. However I was going for low-level implementation to avoid unnecessary allocations. If you think it's not necessary here I can reimplement it differently.

jakubjanecek replied:
So how about now?

jakubjanecek commented:
@Wosin Will you have another look at this please?

devsprint commented:
@jakubjanecek would you have time to complete it during this week?

jakubjanecek commented:
@devsprint It was ready to be merged but merge conflicts appeared.

I think I've fixed everything and rebased onto `master` but some unrelated test is failing. I'd like this to be merged so that I don't have to fix merge conflicts again.

devsprint commented:
@jakubjanecek I will take a look into it as the master or the other PRs that open are not failing during tests.

devsprint commented:
That's a flaky tests. I will fix in a separate PR.

devsprint commented:
@jdegoes can you have a look?

devsprint commented:
Closed in favor of #1834
